hibernate-core/src/main/java/org/hibernate/cfg/IndexOrUniqueKeySecondPass.java

```diff
@@ -72,7 +72,7 @@ public void doSecondPass(Map persistentClasses) throws MappingException {
 		if ( column != null ) {
 			this.table = column.getTable();
 
-			PersistentClass persistentClass = (PersistentClass) persistentClasses.get( column.getPropertyHolder().getEntityName() );
+			PersistentClass persistentClass = (PersistentClass) persistentClasses.get( column.getPropertyHolder().getEntityOwnerClassName() );
 			Property property = persistentClass.getProperty( column.getPropertyName() );
```

Comment:
not sure but probably getting the class name instead of the entity name may cause problems. 
what about https://github.com/hibernate/hibernate-orm/pull/1983/files?